File: common/src/main/java/org/academy/api/client/input/KeyEvent.java
package org.academy.api.client.input;

import net.neoforged.bus.api.Event;
import net.neoforged.bus.api.ICancellableEvent;

public class KeyEvent extends Event implements ICancellableEvent {
    public int key;
    public int scanCode;
    public int action;
    public int modifiers;

    public KeyEvent(int key, int scanCode, int action, int modifiers) {
        this.key = key;
        this.scanCode = scanCode;
        this.action = action;
        this.modifiers = modifiers;
    }
}

File: mod/src/main/kotlin/org/academy/api/client/gui/event/KeyEvent.kt
package org.academy.api.client.gui.event

import com.mojang.blaze3d.platform.InputConstants
import net.minecraft.client.input.InputQuirks

class KeyEvent(type: EventType, keyCode: Int, scanCode: Int, modifiers: Int) : InputEvent(type) {
    val keyCode: Int
    val scanCode: Int
    val modifiers: Int

    init {
        require(!(type != EventType.KEY_PRESSED && type != EventType.KEY_RELEASED)) { "This constructor is for KEY_PRESSED or KEY_RELEASED events." }
        this.keyCode = keyCode
        this.scanCode = scanCode
        this.modifiers = modifiers
    }

    fun hasAltDown(): Boolean {
        return (this.modifiers and InputConstants.MOD_ALT) != 0
    }

    fun hasShiftDown(): Boolean {
        return (this.modifiers and InputConstants.MOD_SHIFT) != 0
    }

    fun hasControlDown(): Boolean {
        return (this.modifiers and InputConstants.MOD_CONTROL) != 0
    }

    fun hasControlDownWithQuirk(): Boolean {
        return (this.modifiers and InputQuirks.EDIT_SHORTCUT_KEY_MODIFIER) != 0
    }
}
